central/src/gitlab.rs

```diff
@@ -0,0 +1,105 @@
+use beam_lib::reqwest::Url;
+use clap::Parser;
+use serde::Deserialize;
+use serde_json::json;
+use shared::{GitLabProjectAccessTokenConfig, SecretResult};
+
+#[derive(Parser)]
+struct GitLabApiConfig {
+    /// The base URL for API calls including the trailing slash, e.g. "https://gitlab.com/"
+    #[clap(long, env)]
+    pub gitlab_url: Url,
+    /// A long-living personal (or impersonation) access token that is used to create short-living project access tokens. Requires at least the "api" scope. Note that group access tokens and project access tokens cannot be used to create project access tokens.
+    #[clap(long, env)]
+    pub gitlab_api_access_token: String,
+}
+
+#[derive(Deserialize)]
+struct GitLabApiReponseBody {
+    token: String,
+}
+
+pub struct GitLabProjectAccessTokenProvider {
+    config: GitLabApiConfig,
+    client: reqwest::Client,
+}
+
+impl GitLabProjectAccessTokenProvider {
+    pub fn try_init() -> Option<Self> {
+        match GitLabApiConfig::try_parse() {
+            Ok(config) => Some(Self {
+                config,
+                client: reqwest::Client::new(),
+            }),
+            Err(e) => {
+                println!("{e}");
+                None
+            }
+        }
+    }
+
+    /// Create a project access token using the GitLab API
+    pub async fn create_token(
+        &self,
+        name: &str,
+        client_config: GitLabProjectAccessTokenConfig,
+    ) -> Result<SecretResult, String> {
+        // Expire in 1 week
+        let expires_at = (chrono::Local::now() + chrono::TimeDelta::weeks(1))
+            .format("%Y-%m-%d")
+            .to_string();
+
+        let response = self
+            .client
+            .post(format!(
+                "{}api/v4/projects/{}/access_tokens",
+                self.config.gitlab_url,
+                urlencoding::encode(&client_config.project_path)
+            ))
```

Comment:
If the user forgets to put a trailing slash at the `gitlab_url` param, this will not only fail but also POST the gitlab admin credentials to some other server (probably non-existing but an attack vector still). Since `gitlab_url` is a URL anyway, may I suggest building the URL using the [URL crate's join function](https://docs.rs/url/latest/url/struct.Url.html#method.join) for a more robust approach.